homeassistant/components/drop_connect/switch.py

```diff
@@ -0,0 +1,128 @@
+    DROPSwitchEntityDescription(
+        key=WATER_SWITCH,
+        translation_key=WATER_SWITCH,
+        icon="mdi:valve",
+        value_fn=lambda device: device.drop_api.water(),
+        set_fn=lambda device, value: device.set_water(value),
+    ),
+    DROPSwitchEntityDescription(
+        key=BYPASS_SWITCH,
+        translation_key=BYPASS_SWITCH,
+        icon="mdi:valve",
+        value_fn=lambda device: device.drop_api.bypass(),
+        set_fn=lambda device, value: device.set_bypass(value),
+    ),
+]
+
+# Defines which switches are used by each device type
+DEVICE_SWITCHES: dict[str, list[str]] = {
+    DEV_FILTER: [BYPASS_SWITCH],
+    DEV_HUB: [WATER_SWITCH, BYPASS_SWITCH],
+    DEV_PROTECTION_VALVE: [WATER_SWITCH],
+    DEV_SOFTENER: [BYPASS_SWITCH],
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the DROP switches from config entry."""
+    _LOGGER.debug(
+        "Set up switch for device type %s with entry_id is %s",
+        config_entry.data[CONF_DEVICE_TYPE],
+        config_entry.entry_id,
+    )
+
+    if config_entry.data[CONF_DEVICE_TYPE] in DEVICE_SWITCHES:
+        async_add_entities(
+            DROPSwitch(hass.data[DOMAIN][config_entry.entry_id], switch)
+            for switch in SWITCHES
+            if switch.key in DEVICE_SWITCHES[config_entry.data[CONF_DEVICE_TYPE]]
+        )
+
+
+class DROPSwitch(DROPEntity, SwitchEntity):
+    """Representation of a DROP switch."""
+
+    entity_description: DROPSwitchEntityDescription
+
+    def __init__(
+        self,
+        coordinator: DROPDeviceDataUpdateCoordinator,
+        entity_description: DROPSwitchEntityDescription,
+    ) -> None:
+        """Initialize the switch."""
+        super().__init__(entity_description.key, coordinator)
+        self.entity_description = entity_description
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return the state of the binary sensor."""
+        switch_value: int | None = self.entity_description.value_fn(self.coordinator)
+        if switch_value == 1:
+            return True
+        if switch_value == 0:
+            return False
+        return None
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn switch on."""
+        await self.entity_description.set_fn(self.coordinator, 1)
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn switch off."""
+        await self.entity_description.set_fn(self.coordinator, 0)
+
+    @property
+    def icon(self) -> str:
+        """Return the icon to use for dynamic states."""
+        _icon = self.entity_description.icon or ""
+        if self.entity_description.key in (WATER_SWITCH, BYPASS_SWITCH):
+            if self.is_on:
+                _icon = "mdi:valve-open"
+            else:
+                _icon = "mdi:valve-closed"
+        return _icon
```

Comment:
Please handle this from the coordinator update that is called instead. The code could be moved to the `DROPEntity` base entity class Here you can set `_attr_icon`, make constants for the icons. For binary sensors you could make a mapping.

Reply:
@jbouwh  Sorry, I don't think I'm understanding this.  Are you saying I should create a new switch subclass in entity.py and move the icon property there?  Something like this:

```python
class DROPSwitchEntity(DROPEntity, SwitchEntity):
    """Representation of a DROP switch device entity."""

    @property
    def icon(self) -> str:
        """Return the icon to use for dynamic states."""
        if self.is_on:
            return ICON_VALVE_OPEN
        return ICON_VALVE_CLOSED
```

... and then use `DROPSwitchEntity` in switch.py?  Would this also apply to the other properties (such as `is_on()`) in switch.py?  You also mentioned binary sensors; do you want me to refactor the (already merged) sensors and binary sensors as part of this PR as well?

I guess I don't understand why it is a problem to have the `icon` property in switch.py.  It appears the most integrations do it that way and since it is using `self.is_on` to determine state, it is already changing based on updates from the DROP hub via the API.  It seems like a lot of unnecessary complexity, and I feel like switch properties belong in switch.py.

Reply:
Left some suggestion on your last commit. Lets use `DROPSwitchEntity` as you initially implemented.